### PR TITLE
Handle GEPs with negative offset and 'nuw' attribute

### PR DESCRIPTION
**Pull request by pkarveti**
This commit fixes incorrect DAG replacement when the 'nuw' attribute
 is added to inbound GEPs with a negative offset.

- If the offset is negative and the 'nuw' attribute is present, the 'nuw' flag is set to false.

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp [hunk 1]
```diff
@@ -4363,10 +4363,15 @@ void SelectionDAGBuilder::visitGetElementPtr(const User &I) {
...
         // interpreted as signed, assume there is no unsigned overflow.
         SDNodeFlags Flags;
-        if (NW.hasNoUnsignedWrap() ||
-            (Offs.isNonNegative() && NW.hasNoUnsignedSignedWrap()))
+        if (NW.hasNoUnsignedWrap()) {
+          if (!Offs.isNonNegative() && NW.hasNoUnsignedSignedWrap()) {
+            Flags.setNoUnsignedWrap(false);
+          } else {
+            Flags.setNoUnsignedWrap(true);
+          }
```
Comment (RKSimon) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:4371]: Merge this if-else into:
`Flags.setNoUnsignedWrap(Offs.isNonNegative() || !NW.hasNoUnsignedSignedWrap()` ?

**Review by nikic — CHANGES_REQUESTED** [review]
This change is incorrect, you are trying to work around an issue somewhere else, but your PR description doesn't have enough information to determine where.

`gep inbounds nuw` with a negative offset returns a poison value, and you should *not* try to suppress that fact.

**Comment by nikic** [comment]
@pkarveti Please check whether your code is clean under `-fsanitize=undefined`. If you want to use a negative index on an array, you need to do so using a signed index, not an unsigned one.

**Comment by iajbar** [comment]
Pavani checked if the code is clean under -fsanitize=undefined and an overflow was detected in the test. This patch can be abandoned.

**Pull request closed without merging**